File: data structures and algorithms/cpp/08_041.cpp
#include <vector>
#include <algorithm>

int minProcessingSpeed(std::vector<int>& batches, int h) {
    int left = 1;
    int right = *std::max_element(
        batches.begin(), 
        batches.end()
    );
    
    auto canComplete = [&](int speed) {
        int hours = 0;
        for (int batch : batches) {
            // Ceiling division
            hours += (batch + speed - 1) / speed;
        }
        return hours <= h;
    };
    
    while (left < right) {
        int mid = left + (right - left) / 2;
        if (canComplete(mid)) {
            right = mid;
        } else {
            left = mid + 1;
        }
    }
    
    return left;
}

File: data structures and algorithms/python/08_041.py
def min_processing_speed(batches, h):

    def can_complete(speed):
        hours = 0
        for batch in batches:
            # Ceiling division to calculate hours needed
            hours += (batch + speed - 1) // speed
        return hours <= h
    
    left, right = 1, max(batches)
    
    while left < right:
        mid = left + (right - left) // 2
        if can_complete(mid):
            right = mid
        else:
            left = mid + 1
            
    return left